### PR TITLE
Using web driver wait

### PR DESCRIPTION
**Pull request by PandaMagnus**
Chris F. found that the expected way of using WebDriverWait is _not_ the way commonly demonstrated online. Swapped out IntelliWait calls for WebDriverWaits with custom exception ignores and delegates passed in. Also sucking in extension project classes as it became clear that the extensions were an integral part of this. In practice, people either wanted it all, or none of it. Extension can be re-broken out later if there's a compelling use case for it.

### CODE REVIEW

**Comment by PandaMagnus** [comment]
Changes for Issue #31

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs [hunk 1]
```diff
@@ -45,9 +46,9 @@ public static void ScrollIntoView(this IWebElement element, IWebDriver driver, i
         /// then tabs out of the field or throws after a certain amount of time
         /// </summary>
         /// <param name="value">Value to send to the element</param>
-        public static async Task FillInWithAndTabWhenReady(this IWebElement element, string value)
+        public static /*async*/ void FillInWithAndTabWhenReady(this IWebElement element, string value)
```
Comment (Keboo) [inline IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs]: I would remove the commented `async`

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs [hunk 2]
```diff
@@ -58,34 +59,52 @@ public static async Task FillInWithAndTabWhenReady(this IWebElement element, str
...
-                await Wait.Until<NoSuchElementException, 
-                    ElementNotVisibleException, 
-                    ElementNotInteractableException,
-                    StaleElementReferenceException>(
-                    () => element.SendKeys(value), TimeSpan.FromSeconds(secondstoTry));
-                count++;
+                wait.Until( c =>
+                {
+                    element.Clear();
+                    return true;
```
Comment (Keboo) [inline IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs]: Why these return values? `WebDriver.Until` appears to have an `Action` overload so you should not need to have a return value unless you intend to use it somewhere.

File: IntelliTect.TestTools.Selenate/GoogleSearch/GoogleBrowser.cs [hunk 3]
```diff
@@ -10,30 +10,24 @@ public class GoogleBrowser : Browser
...
         }
-        
-        public IWebElement FindElement(By by)
-        {
-            return Find(by).ConfigureAwait(false).GetAwaiter().GetResult();
-        }
 
-        // Instead of doing this, I should just do a separate project demonstrating async usage on something more complex than a Google search
-        public async Task<IWebElement> FindElementAsync(By by)
+        public IWebElement FindElement(By by)
```
Comment (kamonson) [inline IntelliTect.TestTools.Selenate/GoogleSearch/GoogleBrowser.cs]: @PandaMagnus have you considered overriding the By element?

Comment (PandaMagnus) [reply IntelliTect.TestTools.Selenate/GoogleSearch/GoogleBrowser.cs]: I already implemented my own IWebElement once before. It did not turn out well. :laughing:

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/WebElementExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            element.SendKeys(Keys.Tab);
+        }
+
+        /// <summary>
+        /// Waits for the element to be in a valid state, 
+        /// then clears the current text and uses SendKeys to send the specified value to the element 
+        /// or throws after a certain amount of time
+        /// </summary>
+        /// <param name="value">Value to send to the element</param>
+        public static void FillInWithWhenReady(this IWebElement element, IWebDriver driver, string value, int secondsToTry = 5)
```
[the remark this answers is not in the document]
Comment (PandaMagnus) [reply IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate/WebElementExtensions.cs]: I'll do a bit more research, but primarily I don't want the code to modify Selenium behavior, outside of retrying should the call fail. I did, however, rename the methods to SendKeysWhenReady

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs (hunk 1 above)
Comment (PandaMagnus) [reply IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs]: I'm going to revert these changes entirely and leave the project as-is, in preparation for deprecating it. In working with Kyle at a client, we've stumbled on a few key things that make it not make sense to have the extensions in a separate project (at least not these ones.) In fact, a couple of these extensions may even get moved into Browser as functions that you pass the element into.

TL;DR, reverting all changes in this class.

File: IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs (hunk 2 above)
Comment (PandaMagnus) [reply IntelliTect.TestTools.Selenate/IntelliTect.TestTools.SelenateExtensions/WebElementExtensions.cs]: The only one I see is for a Func<IWebDriver, bool> argument?

This is the class that WebDriverWait implements:
https://github.com/SeleniumHQ/selenium/blob/master/dotnet/src/webdriver/Support/DefaultWait%7BT%7D.cs
